### PR TITLE
Expose DSL Proof Rules.

### PR DESCRIPTION
**Pull request by abdoo8080**
This PR implements one approach to exposing DSL Proof rules to the user: by making the DSL rule enum public.

Notes:
1. The safe `toString` method is not included in `safe_print.h` because it's generated by the RARE parser.
2. To test the `hash` and `toString`, for all enum values, a "last" enum value is needed. However, the current last enum value (`DISTINCT_BINARY_ELIM`) used in the test is generated by the RARE parser and may change in the future. Potential fixes include: adding `LAST` enum value, moving `NONE` to the end, or defining a `LAST` whose value is generated by the RARE parser.

### CODE REVIEW

**Comment by ajreynol** [comment]
We will need to merge https://github.com/cvc5/cvc5/pull/10614 first.

**Review by ajreynol — CHANGES_REQUESTED** [review]
How about `ProofRewriteRule`?

Feel free to rename other internal names in this PR if this conflicts.

Also, I think we should put the new enum inside of `cvc5_proof_rule.h`, since both pertain to proofs. Is this possible?  What do you think?

**Comment by abdoo8080** [comment]
> How about `ProofRewriteRule`?
> 
> Feel free to rename other internal names in this PR if this conflicts.
> 
> Also, I think we should put the new enum inside of `cvc5_proof_rule.h`, since both pertain to proofs. Is this possible? What do you think?

Works for me! This name does not have any conflicts either, `ProofRewriteRule` vs. `RewriteProofRule` :)

It's possible to move the enums in this file to `cvc5_proof_rule.h`. I don't have a strong opinion, but I prefer to have them separated. There are minor benefits to keep them separate: less files to recompile and ease of formatting and documentation.

**Comment by ajreynol** [comment]
I think we should probably prioritize user convenience in this case. I think it is better for a user have access to all proof rules (core or rewrite) in one header file.

File: include/cvc5/cvc5_proof_rule.h [hunk 1]
```diff
@@ -2266,9 +2266,332 @@ enum ENUM(ProofRule) : uint32_t
...
+ * of the rules are internal ad-hoc rewrites, while others are rewrites
+ * specified by the RARE DSL. This enumeration is used as the first argument to
+ * the :cpp:enumerator:`DSL_REWRITE <cvc5::ProofRule::DSL_REWRITE>` proof rule.
+ */
+enum ENUM(ProofRewriteRule) : uint32_t
+{
+  EVALUE(NONE),
+  // RARE rules
+  // ${rules}$
+  EVALUE(ARITH_PLUS_ZERO),
```
Comment (ajreynol) [inline include/cvc5/cvc5_proof_rule.h:2285]: This is somewhat minor, is it possible to generate a one line comment for each rule here?
Something like
/** Auto-generated from RARE rule arith-plus-zero */

**Review by ajreynol — CHANGES_REQUESTED** [review]

**Review by ajreynol — APPROVED** [review]
LGTM

**Pull request merged**